File: 소마/평범한배낭.py
# 전형적인 knapsack 이차원 dp 배열을 만들어서 해결한다
# 시간 복잡도 O(N * Weight )
# 공간 복잡도 O(N * Weight )인데 O(Weight)로 줄일 수 있긴함.

from sys import stdin

N,W = map(int,stdin.readline().rstrip().split())
bags=[0]
vals=[0]

for _ in range(N):
  b,v =map(int,stdin.readline().rstrip().split())
  bags.append(b)
  vals.append(v)


dp =[[0 for _ in range(W+1)] for _ in range(N+1)]

for n in range(N+1):
  for w in range(W+1):
    
    if w-bags[n] >=0:
      dp[n][w] = max(dp[n-1][w-bags[n]] + vals[n], dp[n-1][w])
    else:
      dp[n][w] = dp[n-1][w]

print(dp[N][W])

File: 소마/평범한배낭.js
const [N, weight] = input.shift().split(" ").map(Number);
const weights = [0];
const vals = [0];
for (let i = 0; i < N; i++) {
  const [w, v] = input[i].split(" ").map(Number);
  weights.push(w);
  vals.push(v);
}

const dp = Array.from({ length: N + 1 }, () =>
  Array.from({ length: weight + 1 }, () => 0)
);

for (let n = 1; n <= N; n++) {
  for (let j = 1; j <= weight; j++) {
    dp[n][j] = Math.max(dp[n - 1][j - weights[n]] + vals[n] || 0, dp[n - 1][j]);
  }
}

console.log(dp[N][weight]);
